File: python/ocrowley_osint/ocrowley_osint/policy.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class OsintAuth:
    actor_id: str
    roles: tuple[str, ...]
    authorization_ref: str
    purpose: str
    lawful_use_acknowledged: bool = False


_OPERATOR_ROLES = frozenset({"osint-operator", "investigator", "counsel"})


def evaluate_osint_auth(auth: OsintAuth, *, require_lawful_ack: bool = False) -> tuple[bool, str]:
    if not auth.authorization_ref.strip():
        return False, "OSINT requires a non-empty authorization_ref"
    if require_lawful_ack and not auth.lawful_use_acknowledged:
        return False, "Caller must acknowledge lawful, proportionate use"
    if not any(r in _OPERATOR_ROLES for r in auth.roles):
        return False, "Requires osint-operator, investigator, or counsel role"
    return True, "authorized"


def assert_osint_allowed(auth: OsintAuth, *, require_lawful_ack: bool = False) -> None:
    ok, reason = evaluate_osint_auth(auth, require_lawful_ack=require_lawful_ack)
    if not ok:
        raise PermissionError(reason)

File: packages/darkweb/src/policy.ts
/**
 * Authorization gates for dark-web index / breach queries.
 */

import { PolicyEngine, type PolicyContext, type PolicyDecision, type PolicyRule } from '@ocrowley/policy';

export type DarkwebAction = 'monitor.entities' | 'search.ahmia' | 'breach.check' | 'intelx.search';

export interface DarkwebAuthorization {
  actorId: string;
  roles: string[];
  authorizationRef: string;
  purpose: string;
  environment?: PolicyContext['environment'];
  /** Explicit acknowledgement that query is lawful and proportionate. */
  lawfulUseAcknowledged: boolean;
}

export function darkwebPolicyRules(): PolicyRule[] {
  return [
    {
      id: 'deny-missing-authorization-ref',
      description: 'Dark-web / breach queries require a non-empty authorization reference.',
      priority: 100,
      effect: 'deny',
      matches: c => !String(c.attributes?.authorizationRef ?? '').trim(),
    },
    {
      id: 'deny-without-lawful-use-ack',
      description: 'Caller must acknowledge lawful, proportionate use.',
      priority: 95,
      effect: 'deny',
      matches: c => c.attributes?.lawfulUseAcknowledged !== true,
    },
    {
      id: 'deny-without-operator-role',
      description: 'Requires osint-operator, investigator, or counsel role.',
      priority: 90,
      effect: 'deny',
      matches: c => !c.roles.some(r => ['osint-operator', 'investigator', 'counsel'].includes(r)),
    },
    {
      id: 'allow-authorized-monitor',
      description: 'Authorized operators may run clearnet index and breach checks.',
      priority: 10,
      effect: 'allow',
      matches: c =>
        Boolean(String(c.attributes?.authorizationRef ?? '').trim()) &&
        c.attributes?.lawfulUseAcknowledged === true &&
        c.roles.some(r => ['osint-operator', 'investigator', 'counsel'].includes(r)),
    },
  ];
}

export function evaluateDarkwebPolicy(action: DarkwebAction, auth: DarkwebAuthorization): PolicyDecision {
  const engine = new PolicyEngine(darkwebPolicyRules());
  const context: PolicyContext = {
    actorId: auth.actorId,
    roles: auth.roles,
    capability: 'darkweb',
    action,
    environment: auth.environment ?? 'development',
    resourceClassification: 'restricted',
    attributes: {
      authorizationRef: auth.authorizationRef,
      purpose: auth.purpose,
      lawfulUseAcknowledged: auth.lawfulUseAcknowledged,
    },
  };
  return engine.evaluate(context);
}

export function assertDarkwebAllowed(action: DarkwebAction, auth: DarkwebAuthorization): void {
  const decision = evaluateDarkwebPolicy(action, auth);
  if (!decision.allowed) {
    throw new Error(`Darkweb policy denied ${action}: ${decision.reason}`);
  }
}
